File: src/main/java/utils/json/JsonStreamFormat.java
package utils.json;

public class JsonStreamFormat {
    private boolean base64, gzipped;
    private StoreType storeType;
    
    public JsonStreamFormat(boolean base64, boolean gzipped, StoreType storeType) {
        this.base64 = base64;
        this.gzipped = gzipped;
        this.storeType = storeType;
    }
    
    public boolean isBase64() {
        return base64;
    }
    
    public boolean isGzipped() {
        return gzipped;
    }
    
    public StoreType getStoreType() {
        return storeType;
    }
    
    public void setBase64(boolean base64) {
        this.base64 = base64;
    }
    
    public void setGzipped(boolean gzipped) {
        this.gzipped = gzipped;
    }
    
    public void setStoreType(StoreType storeType) {
        this.storeType = storeType;
    }
    
    @Override
    public boolean equals(Object o) {
        if (o != null && o instanceof JsonStreamFormat) {
            JsonStreamFormat format = (JsonStreamFormat) o;
            
            return format.base64 == base64 && format.gzipped == gzipped && format.storeType == storeType;
        } else {
            return false;
        }
    }
    
    @Override
    public int hashCode() {
        int hash = (base64)? 8 : 0;
        
        if (gzipped) {
            hash |= 4;
        }
        
        switch (storeType) {
            case NORMAL:
                hash |= 1;
                break;
            case BINARY:
                hash |= 2;
                break;
            case COMPACT:
                hash |= 3;
                break;
        }
        
        return hash;
    }
    
    @Override
    public String toString() {
        return "Format > base64: " + ((base64)? "on" : "off") + ", gzip: " + ((gzipped)? "on" : "off") + 
            ", storeType: " + storeType.desc;
    }
}

File: src/main/kotlin/utils/json/JsonStreamFormat.kt
package utils.json

class JsonStreamFormat(val base64: Boolean, val gzipped: Boolean, val storeType: StoreType) {
    // Auto getter for base64 / gzipped is getBase64/getGzipped
    // These function keep the same api as java implementation
    fun isBase64(): Boolean {
        return base64
    }
    
    fun isGzipped(): Boolean {
        return gzipped
    }

    override fun equals(other: Any?): Boolean {
        if (other != null && other is JsonStreamFormat) {
            return other.base64 == base64 && other.gzipped == gzipped && other.storeType == storeType
        } else {
            return false
        }
    }
    
    override fun hashCode(): Int {
        var hash = if (base64) 8 else 0
        
        if (gzipped) {
            hash = hash.or(4)
        }
        
        when (storeType) {
            StoreType.NORMAL -> {
                hash = hash.or(1)
            }
            StoreType.BINARY -> {
                hash = hash.or(2)
            }
            StoreType.COMPACT -> {
                hash = hash.or(3)
            }
        }
        
        return hash
    }
    
    override fun toString(): String {
        return "Format > base64: ${if (base64) "on" else "off"}, gzip: ${if (gzipped) "on" else "off"}, storeType: $storeType"
    }
}
